quickwit-metastore/src/metastore/single_file_metastore.rs

```diff
@@ -0,0 +1,1352 @@
+/*
+    Quickwit
+    Copyright (C) 2021 Quickwit Inc.
+    Quickwit is offered under the AGPL v3.0 and as commercial software.
+    For commercial licensing, contact us at hello@quickwit.io.
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::collections::HashMap;
+use std::ops::Range;
+use std::path::Path;
+use std::sync::{Arc, RwLock};
+
+use async_trait::async_trait;
+use quickwit_doc_mapping::DocMapping;
+use quickwit_storage::{PutPayload, Storage};
+
+use crate::{
+    IndexMetadata, IndexUri, MetadataSet, Metastore, MetastoreErrorKind, MetastoreResult, SplitId,
+    SplitMetadata, SplitState, FILE_FORMAT_VERSION,
+};
+
+/// Single file meta store implementation.
+pub struct SingleFileMetastore {
+    storage: Arc<dyn Storage>,
+    data: Arc<RwLock<HashMap<IndexUri, MetadataSet>>>,
+}
+
+#[allow(dead_code)]
+impl SingleFileMetastore {
+    /// Creates a meta store given a storage.
+    pub async fn new(storage: Arc<dyn Storage>) -> MetastoreResult<Self> {
+        Ok(SingleFileMetastore {
+            storage,
+            data: Arc::new(RwLock::new(HashMap::new())),
+        })
+    }
+}
+
+#[async_trait]
+impl Metastore for SingleFileMetastore {
+    async fn create_index(
+        &self,
+        index_uri: IndexUri,
+        _doc_mapping: DocMapping,
+    ) -> MetastoreResult<()> {
+        // Serialized metadata set.
+        let contents: Vec<u8>;
+
+        {
+            let mut data = self.data.write().unwrap();
+
+            // Check for the existence of index.
+            if data.contains_key(&index_uri) {
```

Comment:
There is a recurring pattern here.

Two options to make the code simpler.

# Option 1

Introduce a private fn in the metastore...
```
fn mutate_meta_store<F: FnOnce(Entry<_,_,_>) -> Result<_, _>>
```

# Option 2 (my personal favorite)

Introduce a get and a put method that individually take the lock.
The lock will be taken twice (get as a read, and put as a write, and get would have to clone the metadata).
(Note the clone part might not be great for performance if we have a millions of splits.)

What I like about this solution is that the code is trivial and cannot panic while the lock is being held.

Reply:
It might not be as easy as I suggested...